packages/@aws-cdk/aws-iam/lib/policy-document.ts

```diff
@@ -197,12 +197,20 @@ export class PolicyStatement extends Token {
    * Indicates if this permission has a "Principal" section.
    */
   public get hasPrincipal() {
-    return this.principal && this.principal.length > 0;
+    return Object.keys(this.principal).length > 0;
   }
 
   public addPrincipal(principal: PolicyPrincipal): PolicyStatement {
     const fragment = principal.policyFragment();
-    this.principal.push(fragment.principalJson);
+    for (const key of Object.keys(fragment.principalJson)) {
```

Comment:
I'm not sure this is correct in general. I believe the '{ AWS: ... }` part of a principal is a sort of type tag, and merging them does not make sense.

For example, merging the following two principals:

```
{ AWS: '123456789012' }
{ GCP: 'booh' }  // even though this doesn't exist
```

to come up with:

```
{
  AWS: '123456789012',
  GCP: 'booh' 
}
```

Is NOT a valid principal. 

The correct behavior would be to merge if the one and only key in the principal is the same, and to throw an error otherwise.

Reply:
And if your argument is, "but AWS is the only key that can exist anyway", then don't generally merge keys, just check for the existence of an "AWS" key.

Reply:
My argument is that the key can be one of `AWS`, `CanonicalUser`, `Federated` or `Service` (currently). As far as [the documentation](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_principal.html) tells, it *cannot* be an array anyway, so the new code is in any case more correct than what was before.

The alternative here is to *require* there be only exactly one key, and restrict adding more principals to only using the same key. I can do this if that makes you feel any better?